Detail the impacts of the new Symfony layout

jolelievre opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop developer documentation! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines on how to contribute:
https://devdocs.prestashop-project.org/8/contribute/documentation/how/
------------------------------------------------------------------------------>

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 9.x
| Description?  | Detail the impacts of the new Symfony layout
| Fixed ticket? | ~

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

jolelievre committed:
Detail the impacts of the new Symfony layout

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -117,6 +117,95 @@ Display event listeners for the back office in `dev` environment (default value
+- employee dropdown
+- the multistore header (when multistore is enabled)
+- the page toolbar (which includes breadcrumb and top action buttons)
+- the footer (which only contains a displayBackOfficeFooter display hook)
```
```suggestion
All the back office pages share a common layout, composed of a few elements:
- the `<head>` element that includes all the CSS and JS (among other things)
- the side navigation menu
- the header of the page, which is itself composed of:
-- quick accesses
-- search form
-- notifications center
-- employee dropdown
-- the multistore header (when multistore is enabled)
-- the page toolbar (which includes breadcrumb and top action buttons)
- the footer (which only contains a `displayBackOfficeFooter` hook)
```

formatting

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -117,6 +117,95 @@ Display event listeners for the back office in `dev` environment (default value
+- the page toolbar (which includes breadcrumb and top action buttons)
+- the footer (which only contains a displayBackOfficeFooter display hook)
+
+Until PrestaShop 8.1 all these common elements were handled by legacy code, so even on migrated page there was always a background legacy controller based on `AdminController` in charge of building the layout data and rendering it. Once the layout was rendered the central content of Symfony pages was included in the middle of it. It means no page was completely free from the legacy controllers wh ...
```
```suggestion
Until PrestaShop 8.1, all these common elements were handled by legacy code, so even on the migrated page, there was always a background legacy controller based on `AdminController` in charge of building the layout data and rendering it. Once the layout was rendered the central content of Symfony pages was included in the middle of it. It means no page was completely free from the legacy controllers, which would ultimately block the end of migration, while they are intended to disappear completely.
```

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -117,6 +117,95 @@ Display event listeners for the back office in `dev` environment (default value
+
+Until PrestaShop 8.1 all these common elements were handled by legacy code, so even on migrated page there was always a background legacy controller based on `AdminController` in charge of building the layout data and rendering it. Once the layout was rendered the central content of Symfony pages was included in the middle of it. It means no page was completely free from the legacy controllers wh ...
+
+In PrestaShop 9.0 all this layout part is now fully handled by Symfony, we use <a href="https://symfony.com/bundles/ux-twig-component/current/index.html" target="_blank">Twig components</a> to render each element independently. The code is, therefore, easier to understand, a component class is responsible for fetching/building the data while the actual rendering is based on Twig. See our <a href= ...
```
```suggestion
In PrestaShop 9.0 all this layout part is now fully handled by Symfony, we use <a href="https://symfony.com/bundles/ux-twig-component/current/index.html" target="_blank">Twig components</a> to render each element independently. The code is, therefore, easier to understand, a component class is responsible for fetching/building the data, while the actual rendering is based on Twig. See our <a href="https://github.com/PrestaShop/PrestaShop/tree/develop/src/PrestaShopBundle/Twig/Component" target="_blank">Layout components</a> for more details.
```

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -117,6 +117,95 @@ Display event listeners for the back office in `dev` environment (default value
+
+In PrestaShop 9.0 all this layout part is now fully handled by Symfony, we use <a href="https://symfony.com/bundles/ux-twig-component/current/index.html" target="_blank">Twig components</a> to render each element independently. The code is, therefore, easier to understand, a component class is responsible for fetching/building the data while the actual rendering is based on Twig. See our <a href= ...
+
+In legacy pages the same principle is used, Symfony is now in charge of rendering all the layout, and we use some <a href="https://github.com/PrestaShop/PrestaShop/tree/develop/src/PrestaShopBundle/Twig/Component/Legacy" target="_blank">Legacy layout components</a> that follow the same architecture but render different twig templates to fit with the old **default theme** (that is based on old boo ...
```
```suggestion
On legacy pages, we follow the same principle. Symfony is now in charge of rendering all the layout, and we use some <a href="https://github.com/PrestaShop/PrestaShop/tree/develop/src/PrestaShopBundle/Twig/Component/Legacy" target="_blank">Legacy layout components</a> that follow the same architecture, but render different Twig templates to fit with the old **default theme** (that is based on old Bootstrap and includes old legacy helpers like `HelperForm` and `HelperList`).
```

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -117,6 +117,95 @@ Display event listeners for the back office in `dev` environment (default value
+
+In legacy pages the same principle is used, Symfony is now in charge of rendering all the layout, and we use some <a href="https://github.com/PrestaShop/PrestaShop/tree/develop/src/PrestaShopBundle/Twig/Component/Legacy" target="_blank">Legacy layout components</a> that follow the same architecture but render different twig templates to fit with the old **default theme** (that is based on old boo ...
+
+**So what changes for my module pages?**
```
```suggestion
**What changes for my module pages?**
```

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -117,6 +117,95 @@ Display event listeners for the back office in `dev` environment (default value
+
+**So what changes for my module pages?**
+
+We refactored this layout with maximum backward compatibility in mind, the HTML layout itself had minimum changes (in both migrated and legacy pages), we tried to keep the same hooks in both the PHP code and the twig templates. We even introduced a fake legacy controller in migrated pages that has no logic but is kept mostly as a DTO to contain things like CSS files and JS files, because that's w ...
```
```suggestion
We refactored this layout with maximum backward compatibility in mind. The HTML layout itself had minimum changes (in both migrated and legacy pages). We tried to keep the same hooks in both the PHP code and the Twig templates. We even introduced a fake legacy controller in migrated pages that has no logic but is kept mostly as a [DTO](https://en.wikipedia.org/wiki/Data_transfer_object) to contain things like CSS files and JS files because that's where most modules add their content.
```

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -117,6 +117,95 @@ Display event listeners for the back office in `dev` environment (default value
+**So what changes for my module pages?**
+
+We refactored this layout with maximum backward compatibility in mind, the HTML layout itself had minimum changes (in both migrated and legacy pages), we tried to keep the same hooks in both the PHP code and the twig templates. We even introduced a fake legacy controller in migrated pages that has no logic but is kept mostly as a DTO to contain things like CSS files and JS files, because that's w ...
+But we still had to change the controller workflow, many code in `AdminController` is now not useful anymore as it is used to initialize the layout variables (now handled by Twig components) some methods also renders or write output directly which we prevent as we need to get the rendered content as a string to correctly integrate it inside the new layout so some internal methods had to be split  ...
```
```suggestion
We still had to change the controller workflow. Many of the functions in `AdminController` are no longer useful as they are used to initialize the layout variables (now handled by Twig components). Some methods also render or write output directly, which we prevent as we need to get the rendered content as a string to integrate it correctly inside the new layout. This means some internal methods had to be split to avoid unwanted usage.
```

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -117,6 +117,95 @@ Display event listeners for the back office in `dev` environment (default value
+
+**Legacy workflow:**
+
+- `Dispatcher::dispatch` would detect the legacy controller and instanciate it
```
```suggestion
- `Dispatcher::dispatch` would detect the legacy controller and instantiate it
```

kpodemski replied:
either legacy workflow is still somewhere, and we should keep "detects" or it's a song of the past, and we should use `would detect`, this is relevant for all other elements on this list

kpodemski replied:
maybe you could link to this page https://devdocs.prestashop-project.org/8/development/architecture/legacy/legacy-controllers/?

jolelievre replied:
It's in the past for the dispatcher, for other methods some are still called So I tried to keep some consistency and use present for methods that are still used

I didn't know about this diagram, it's great!! We should indeed create a similar one for the new layout I didn't know about mermaid but it seems great!

I'd prefer if we add this workflow in another PR though, because I'm on vacation at the end of the week and I'd like to finalize all the text content this week at least.

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -117,6 +117,95 @@ Display event listeners for the back office in `dev` environment (default value
+**Legacy workflow:**
+
+- `Dispatcher::dispatch` would detect the legacy controller and instanciate it
+- `AdminController::run` was executed by the `Dispatcher` which was split into
```
```suggestion
- `AdminController::run` was executed by the `Dispatcher`, which was split into
```

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -117,6 +117,95 @@ Display event listeners for the back office in `dev` environment (default value
+
+- A special routing matching URLs like `?controller=LegacyAdminController` checks if the URL is valid thanks to `LegacyRouterChecker::check`
+  - the query parameter must match an existing legacy controller class
+  - the controller object is instanciated and `AdminController::init` is called, this is important to know if the controller is anonymous, this is the proper moment to define `AdminController::allowAnonymous`
```
```suggestion
  - the controller object is instantiated, and `AdminController::init` is called, this is important to know if the controller is anonymous, this is the proper moment to define `AdminController::allowAnonymous`
```

kpodemski replied:
consider explaining what it means to call a controller "anonymous"

kpodemski replied:
also, we use Mermaid in the docs, so if you want, you could use https://mermaid.live/edit to generate flowchart :D

jolelievre replied:
Yep good idea, but as I said in another comment probably in another PR

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -117,6 +117,95 @@ Display event listeners for the back office in `dev` environment (default value
+      - calls `Cookie::write` which was previously called via `smartyOutputContent`
+      - get the modal content via `AdminController::renderModal`
+
+The two workflows should render the same result, many methods from the legacy workflow are not executed anymore because they lost their purpose, but in case you overrode one of those methods be aware that they are no longer called directly so they are breaking changes:
```
```suggestion
The two workflows should render the same result. Many methods from the legacy workflow are not executed anymore because they lost their purpose, but in case you override one of those methods, be aware that they are no longer called directly, so there are breaking changes:
```

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -117,6 +117,95 @@ Display event listeners for the back office in `dev` environment (default value
+- `AdminController::initFooter`
+- `AdminController:display`
+
+Also be aware that many smarty variables are no longer defined because they were only used to render the layout.
```
```suggestion
Also, be aware that many Smarty variables are no longer defined because they were only used to render the layout.
```

kpodemski commented on modules/core-updates/9.0.md:
```diff
@@ -117,6 +117,95 @@ Display event listeners for the back office in `dev` environment (default value
+- [Clean Smarty variables that were only relevant for the layout internal use](https://github.com/PrestaShop/PrestaShop/pull/34560)
+- [Legacy pages are rendered by Symfony](https://github.com/PrestaShop/PrestaShop/pull/34783)
+
+For more details about the changes you can check the content of the <a href="https://github.com/PrestaShop/PrestaShop/issues/32875" target="_blank">Symfony layout Epic</a>.
```
```suggestion
For more details about the changes, you can check the content of the <a href="https://github.com/PrestaShop/PrestaShop/issues/32875" target="_blank">Symfony layout Epic</a>.
```

kpodemski reviewed (commented):
I added some suggestions.

If you have some time, consider using Mermaid, especially since a live editor that generates the code is available. If not, I'll try to find some time to improve this page afterward.

On my side, I will test some of the legacy modules I have to better understand the impact of these changes.

Thanks @jolelievre 🙏🏻

jolelievre commented:
Thanks @kpodemski 

I think I addressed all your feedback, I agree that there's still room for improvement Especially adding a diagram for the Symfony workflow, but I'd prefer to do it in another PR.

I'm leaving for three weeks and I won't have time to finish it this week.

kpodemski reviewed (approved):
@jolelievre sure thing! enjoy your holidays!
